File: T/42-trapping-rain-water/main.go
package main

import "fmt"

func trap(height []int) int {
	sIdx := -1
	for i := 0; i < len(height)-1; i++ {
		if height[i] > height[i+1] {
			sIdx = i
			break
		}
	}
	if sIdx == -1 || sIdx == len(height)-2 {
		return 0
	}

	sum := 0
	var stack []int
	var offset []int
	for i := sIdx; i < len(height)-1; i++ {
		stack = append(stack, height[i])
		offset = append(offset, i)
		if height[i] < height[i+1] {
			level := 0
			off := -1
			popIdx := -1
			for m := len(stack) - 1; m >= 0; m-- {
				if stack[m] > height[i+1] {
					popIdx = m
					level = height[i+1]
					off = offset[m]
					break
				}
			}
			if popIdx == -1 {
				level = stack[0]
				off = 0
			}
			for m := popIdx + 1; m < len(stack); m++ {
				b := stack[m]
				o := offset[m]
				sum += (level - b) * (i - off)
				level = b
				off = o
			}
			stack = stack[0 : popIdx+1]
			offset = offset[0 : popIdx+1]
		}
	}
	return sum
}

func main() {
	fmt.Println(trap([]int{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}))
	fmt.Println(trap([]int{8, 2, 8, 9, 0, 1, 7, 7, 9}))
}

File: T/42-trapping-rain-water/main.rs
struct Solution {}

impl Solution {
  pub fn trap(height: Vec<i32>) -> i32 {
    let mut stack: Vec<(i32, usize)> = Vec::new();
    let mut sum = 0;
    height.iter().enumerate().for_each(|(idx, v)| {
      if stack.len() == 0 || stack[stack.len() - 1].0 >= *v {
        stack.push((*v, idx));
      } else if stack[stack.len() - 1].0 < *v {
        let mut last = stack.pop().unwrap();
        while stack.len() > 0 && stack[stack.len() - 1].0 < *v {
          let mut prev = stack.pop().unwrap();
          sum += (prev.0 - last.0) * (idx - prev.1 - 1) as i32;
          last = prev;
        }
        if stack.len() > 0 {
          sum += (*v - last.0) * (idx - stack[stack.len() - 1].1 - 1) as i32;
        }

        stack.push((*v, idx));
      }
    });
    sum
  }
}
